pkg/providers/filesystem.go

```diff
@@ -0,0 +1,134 @@
+package providers
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/karrick/godirwalk"
+	"github.com/spectralops/teller/pkg/core"
+
+	"github.com/spectralops/teller/pkg/logging"
+)
+
+type FileSystem struct {
+	logger        logging.Logger
+	rootDirectory string
+}
+
+// NewFileSystem creates new provider instance
+func NewFileSystem(logger logging.Logger) (core.Provider, error) {
+
+	return &FileSystem{
+		logger:        logger,
+		rootDirectory: "",
+	}, nil
+}
+
+// Name return the provider name
+func (f *FileSystem) Name() string {
+	return "FileSystem"
+}
+
+// Put will create a new single entry
+func (f *FileSystem) Put(p core.KeyPath, val string) error {
+	return f.writeFile(f.getFilePath(p.Path), val)
+}
+
+// PutMapping will create a multiple entries
+func (f *FileSystem) PutMapping(p core.KeyPath, m map[string]string) error {
+	for k, v := range m {
+		ap := p.SwitchPath(fmt.Sprintf("%v/%v", f.getFilePath(p.Path), k))
+		err := f.Put(ap, v)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetMapping returns a multiple entries
+func (f *FileSystem) GetMapping(p core.KeyPath) ([]core.EnvEntry, error) {
+
+	findings := []core.EnvEntry{}
+	err := godirwalk.Walk(f.getFilePath(p.Path), &godirwalk.Options{
+		Callback: func(osPathname string, de *godirwalk.Dirent) error {
+			if de.IsRegular() {
+				content, err := f.readFile(osPathname)
+				if err != nil {
+					f.logger.WithError(err).WithField("path", p.Path).Debug("file not found in path")
+					return nil
+				}
+				findings = append(findings, p.FoundWithKey(strings.Replace(osPathname, fmt.Sprintf("%s/", p.Path), "", 1), string(content)))
+			}
+			return nil
+		},
+		Unsorted: true,
+	})
+
+	return findings, err
+}
+
+// Get returns a single entry
+func (f *FileSystem) Get(p core.KeyPath) (*core.EnvEntry, error) {
+	content, err := f.readFile(f.getFilePath(p.Path))
+	if err != nil {
+		f.logger.WithError(err).WithField("path", p.Path).Debug("file not found in path")
+		return nil, err
+	}
+	ent := p.Found(string(content))
+	return &ent, nil
+}
+
+// Delete will delete entry
+func (f *FileSystem) Delete(kp core.KeyPath) error {
+	deletePath := f.getFilePath(kp.Path)
+	fileInfo, err := os.Stat(deletePath)
+	if err != nil {
+		return err
+	}
+	// to make the delete safely, we allow deleting a single file only
+	if fileInfo.IsDir() {
+		return errors.New("delete folder is not supported")
+	}
+
+	return os.Remove(deletePath)
+}
+
+// DeleteMapping will delete the given path recessively
+func (f *FileSystem) DeleteMapping(kp core.KeyPath) error {
+	return fmt.Errorf("provider mapping %s does not implement delete yet", f.Name())
+}
+
+func (f *FileSystem) getFilePath(p string) string {
```

Comment:
What happens if root is `/data`
but user supplies `/root`

Do we get?
1. `/data/root` (good)
2. `/root` (bad)

This depends on how path resolution is implemented in Go.
Worth checking out to avoid surprises.

In other words, it feels that getFilePath should always ensure that path `p` has a prefix `rootDirectory` no matter what.

Reply:
This function aims to unitesting for using the `AssertProvider` function.
The `rootDirectory` parameters is initialized empty (the user can't change it), and we always take what configured  .tellet.yaml